source/distributed/tria_base.cc

```diff
@@ -342,6 +346,201 @@ namespace parallel
+          this->n_global_levels(), 0);
+
+        MPI_Exscan(n_locally_owned_cells.data(),
+                   cell_index.data(),
+                   this->n_global_levels(),
+                   Utilities::MPI::internal::mpi_type_id(
+                     n_locally_owned_cells.data()),
+                   MPI_SUM,
+                   this->mpi_communicator);
+
+        // 3) determine global number of "active" cells on each level
+        std::vector<types::global_cell_index> n_cells_level(
+          this->n_global_levels(), 0);
+
+        MPI_Allreduce(n_locally_owned_cells.data(),
+                      n_cells_level.data(),
+                      this->n_global_levels(),
+                      Utilities::MPI::internal::mpi_type_id(
+                        n_locally_owned_cells.data()),
+                      MPI_SUM,
+                      this->mpi_communicator);
+
+        // 4) give global indices to locally-owned cells on level and mark
+        //    all other cells as invalid
+        for (auto cell : this->cell_iterators())
+          if (cell->level_subdomain_id() == this->locally_owned_subdomain())
+            cell->set_global_level_cell_index(cell_index[cell->level()]++);
+          else
+            cell->set_global_level_cell_index(numbers::invalid_dof_index);
+
+        // 5) update the numbers of ghost level cells
+        GridTools::exchange_cell_data_to_level_ghosts<
+          types::global_cell_index,
+          dealii::Triangulation<dim, spacedim>>(
+          *this,
+          [](const auto &cell) {
+            return cell->global_level_cell_index_local();
+          },
+          [](const auto &cell, const auto &id) {
+            return cell->set_global_level_cell_index(id);
+          });
+
+        number_cache.level_cell_index_partitioners.resize(
+          this->n_global_levels());
+
+        // 6) set up cell partitioners for each level
+        for (unsigned int l = 0; l < this->n_global_levels(); ++l)
+          {
+            IndexSet is_local(n_cells_level[l]);
+            IndexSet is_ghost(n_cells_level[l]);
+
+            for (const auto &cell : this->cell_iterators_on_level(l))
+              if (cell->level_subdomain_id() !=
+                  dealii::numbers::artificial_subdomain_id)
+                {
+                  const auto index = cell->global_level_cell_index_local();
+
+                  if (index == numbers::invalid_dof_index)
+                    continue;
+
+                  if (cell->level_subdomain_id() ==
+                      this->locally_owned_subdomain())
+                    is_local.add_index(index);
+                  else
+                    is_ghost.add_index(index);
+                }
+
+            number_cache.level_cell_index_partitioners[l] =
+              Utilities::MPI::Partitioner(is_local,
+                                          is_ghost,
+                                          this->mpi_communicator);
```

Comment:
Given that we didn't build that information before at all, this is a pretty compact code in relation to the content of information contained here.

Reply:
It has been prepared by a couple of PRs half a year ago ;)